src/appleseed/renderer/kernel/lighting/lighttree.cpp

```diff
@@ -0,0 +1,332 @@
+
+    // Build the light-tree.
+    typedef foundation::bvh::Builder<LightTree, Partitioner> Builder;
+    Builder builder;
+    builder.build<foundation::DefaultWallclockTimer>(*this, partitioner, m_items.size(), 1);
+    statistics.insert_time("build time", builder.get_build_time());
+
+    // Reorder m_items vector to match the ordering in the LightTree.
+    if (!m_items.empty())
+    {
+        m_built = true;
+
+        const std::vector<size_t>& ordering = partitioner.get_item_ordering();
+        assert(m_items.size() == ordering.size());
+
+        // Reorder the items according to the tree ordering.
+        ItemVector temp_lights(ordering.size());
+        foundation::small_item_reorder(
+            &m_items[0],
+            &temp_lights[0],
+            &ordering[0],
+            ordering.size());
+
+        // Set total luminance and level for each node of the LightTree.
+        recursive_node_update(0, 0);
+    }
+    
+    // Print light-tree statistics.
+    RENDERER_LOG_INFO("%s",
+        foundation::StatisticsVector::make(
+            "light-tree statistics",
+            statistics).to_string().c_str());
+
+    RENDERER_LOG_INFO("number of light sources: %zu", m_light_sources.size());
+    RENDERER_LOG_INFO("number of nodes: %zu", m_nodes.size());
+    RENDERER_LOG_INFO("tree depth: %zu", m_tree_depth);
+}
+
+//
+// Control output of the built tree bboxes into the VPython file.
+// The py file is written to the cwd.
+// TODO: Add a possibility to shift each level of bboxes along the z-axis.
+//
+
+void LightTree::draw_tree_structure(
+    const std::string&          filename_base,
+    const foundation::AABB3d&   root_bbox,
+    const bool                  separate_by_levels) const
+{
+    // Vpython tree output
+    const double Width = 0.1;
+
+    if (separate_by_levels)
+    {
+        const char* color = "color.green";
+        for (size_t parent_level = 0; parent_level < m_tree_depth; parent_level++)
+        {
+            const std::string filename = filename_base
+                                       + "_"
+                                       + std::to_string(parent_level + 1)
+                                       + ".py";
+            foundation::VPythonFile file(filename.c_str());
+            file.draw_axes(Width);
+
+            // Draw the initial bbox.
+            file.draw_aabb(root_bbox, color, Width);
+
+            // Find the parent node to draw child bboxes from.
+            for (size_t i = 0; i < m_nodes.size(); i++)
+            {
+                if (m_nodes[i].is_leaf())
+                    continue;
+
+                if (m_nodes[i].get_level() == parent_level)
+                {
+
+                    const auto& bbox_left = m_nodes[i].get_left_bbox();
+                    const auto& bbox_right = m_nodes[i].get_right_bbox();
+
+                    file.draw_aabb(bbox_left, color, Width);
+                    file.draw_aabb(bbox_right, color, Width);
+                }
```

Comment:
Yes, I am not sure why would I need a break statement as I need to check the whole list to find nodes of the given level?